stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/attachment/internal/StreamAttachmentViewFactory.kt

```diff
@@ -0,0 +1,40 @@
+package io.getstream.chat.android.ui.message.list.adapter.viewholder.attachment.internal
+
+import android.content.Context
+import android.view.View
+import android.view.ViewGroup
+import io.getstream.chat.android.client.models.Attachment
+import io.getstream.chat.android.core.internal.InternalStreamChatApi
+import io.getstream.chat.android.ui.common.extensions.internal.hasLink
+import io.getstream.chat.android.ui.message.list.adapter.internal.MessageListItemViewTypeMapper.isMedia
+import io.getstream.chat.android.ui.message.list.adapter.view.internal.FileAttachmentsView
+import io.getstream.chat.android.ui.message.list.adapter.view.internal.LinkAttachmentView
+import io.getstream.chat.android.ui.message.list.adapter.view.internal.MediaAttachmentsGroupView
+import io.getstream.chat.android.ui.message.list.adapter.viewholder.attachment.AttachmentViewFactory
+
+@InternalStreamChatApi
+internal class StreamAttachmentViewFactory : AttachmentViewFactory {
+    override fun createLinkAttachmentView(linkAttachment: Attachment, context: Context): View {
+        require(linkAttachment.hasLink()) { "Can create link view only for attachments with link" }
```

Comment:
Why do links need their own method? Can't we check for whether something is a link in the `createAttachmentsView` method, with the same `hasLink` that we use here?

Reply:
Right now we have simultaneously three modes for attachments:
1) Link + Attachment
2) Media attachments only
3) File attachments only

We define that if we have both attachments and links then we should show attachments on the top and link below. If we had links or other attachments then we could have a single method where we put all attachments that we have and it returns a single view for attachments.

Reply:
Maybe when we'll fuse VHs into one we will be able to make attachment view factory easy. Right now this PR is transit step for attachment customization

Reply:
I'm a bit worried about already adding this as public API as-is, because we won't be able to release a new version until we've followed it up and changed it. But if we expect to move fast with the follow-up that cleans it up, I guess we can proceed like this. Would really like to see [all this logic](https://github.com/GetStream/stream-chat-android/blob/46cc09082c1290287a1b98629705696f00a5f722/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/internal/PlainTextWithFileAttachmentsViewHolder.kt#L100-L129) moved to the factory if that's possible, so that the factory only exposes one method for clients to override.

Alternatively, we could ask design if it's okay to remove link previews if we already have other attachments on the message.